### ISSUE TITLE
[5.5.2]: Issues reading compressed netCDF files with toolsUI 5.5.2 via nj22Config.xml 

### ISSUE THREAD

**Issue by mcucchi9** [Unidata/netcdf-java#997]
### Description

I'm trying to read a JPEG-LS compressed netCDF file with ToolsUI 5.5.2 using a decompression library written in C, by allowing the use of the netCDF-C library for reading purposes through the `$HOME\.unidata\nj22Config.xml` configuration file (https://github.com/Unidata/thredds/issues/1063). 
While I am able to do it (and also perform a decompression through the "Writer" tab functionalities) with v5.1.0, I can't with the latest version: as soon as I open the compressed file I get the following warning message on my terminal:

```
[16:57:42.645 WARN  u.nc2.internal.iosp.hdf5.H5headerNew]   Variable effective_radiance has unknown Filter(s) =    MessageFilter filters=
    Filter id= 32018 flags = 1 nValues=13 name= HDF5 JPEG-LS filter data = 2 1 279 11136 16 1 0 0 0 0 0 0 0 
```

A similar issue applies to Panoply 5.0.2 (based upon netcdfAll-5.5.2) with respect to Panoply 4.10.12 (based upon netcdfAll-5.1.0): while I am able to open and plot variables of JPEG-LS compressed files with v4.10.12 exploiting the `$HOME\.unidata\nj22Config.xml` file, I can't with v5.0.2, getting the following error message when trying to create a plot:

```
There was an exception preparing the data: Unknown filter type=32018
```

### Environmental information

**OS**: Ubuntu 20.04.4 LTS
**Java version**: openjdk version "13.0.7" 2021-04-20

#### Other details

- I'm working in a ``conda`` environment where the decompression filter is correctly configured to be used by the netCDF-C library (tested with netCDF-C tools such as ``nccopy``)
- the ``$HOME/.unidata/nj22Config.xml`` file is configured as follows
   ```
   <nj22Config>
     <Netcdf4Clibrary>
        <libraryPath>$NETCDFC_LIB_PATH</libraryPath>
        <libraryName>netcdf</libraryName>
        <useForReading>true</useForReading>
     </Netcdf4Clibrary>
   </nj22Config>
   ```
   where `$NETCDFC_LIB_PATH` is the explicit full path to the ``conda`` environment ``lib`` directory containing the netCDF-c library `libnetcdf.so`
- I tried to launch the mentioned Java applications both with
   ```
   java -jar <full_path_to_the_java_application_jar>
   ```
   and
   ```
   java -jar <full_path_to_the_java_application_jar> -nj22Config $HOME/.unidata/nj22Config.xml
   ```
   obtaining the same results

**Comment by cofinoa** [comment]
First, is the ncdump, for example, from the conda env working? (you say that it is working, but it's just to confirm)

Second, in the .nj22Config.xml are you writing the actual PATH or are you writing "$NETCDFC_LIB_PATH", expecting to be interpolated as a environment variable?

Note: can you provide a .nc example file?

**Comment by mcucchi9** [comment]
Hi @cofinoa 

> First, is the ncdump, for example, from the conda env working? (you say that it is working, but it's just to confirm)

Yes it is, as both toolsUI 5.1.0 and Panoply 4.10.12 are.

> Second, in the .nj22Config.xml are you writing the actual PATH or are you writing "$NETCDFC_LIB_PATH", expecting to be interpolated as a environment variable?

I'm writing the actual full path.

> Note: can you provide a .nc example file?

You can find a test compressed netCDF file inside the attached ZIP archive.

Thanks

[test_compressed.zip](https://github.com/Unidata/netcdf-java/files/8492210/test_compressed.zip)

**Comment by cofinoa** [comment]
@mcucchi9 

>> First, is the ncdump, for example, from the conda env working? (you say that it is working, but it's just to confirm)

> Yes it is, as both toolsUI 5.1.0 and Panoply 4.10.12 are.

Sorry, I had expressed very bad. What I mean by *ncdump* working, is if you are able to dump the data from encoded variable with the JPEG-LS filter?. For example, my conda env it's not able to dump the data section for the variable `effective_radiance` which is in the group `/data/vis_04/measured`
```
$ ncdump -g "/data/vis_04/measured" -v effective_radiance test_compressed.nc
```
Note: My conda env it's using:
```
hdf5                      1.12.1          nompi_h2750804_101    conda-forge
libnetcdf                 4.8.1           nompi_hb3fd0d9_101    conda-forge
```

which one are you using? from which conda channel?

**Comment by mcucchi9** [comment]
Hi @cofinoa 

> Sorry, I had expressed very bad. What I mean by ncdump working, is if you are able to dump the data from encoded variable with the JPEG-LS filter?

Yes, I got what you meant, and I confirm `ncdump` (as well as toolsUI 5.1.0 and Panoply 4.10.12) is able to read data from variables encoded with the JPEG-LS filter (like the one you tried to dump). 
You are not able to do the same because you do not have the C JPEG-LS decompression library installed and configured in your environment, which unfortunately at the moment I'm not able to share.

Anyway I suspect that the issue here is that somehow, regardless of the particular decompression library, after v5.1.0 netcdfAll lost the ability to use netcdf-c for reading/decompressing purposes, or alternatively to properly use the nj22Config.xml configuration file.

Thanks

**Comment by cofinoa** [comment]
Hi @mcucchi9 , 

No, the 5.5.2 version hasn't lost the ability to use the netcdf-c, as you can see from this screen capture where I have been able to read the encoded variable using the netcdf-c: 

![image](https://user-images.githubusercontent.com/5098603/163572186-16d6c5bc-0df1-4e8f-8844-aff45a60ce43.png)

as you can see the ToolsUI it's able to load and read the variable using the netcdf-c library with the JPEG-LS filter. 

The problem it's with some part of the `nj22Config.xml` and in fact with the IOSP registration of the netcdf-c reader as default IOSP for NC4 files. 

I'm not able to share the current solution, but let me know if you are really interested and I can try a fix.

**Comment by cofinoa** [comment]
@mcucchi9, yes, I can confirm that there is an issue with the IOSP registration. 

Meanwhile a fix it's developed, I think I can develop and provide a patch for the 5.5.2 release. Let me know if you are interested!

**Comment by mcucchi9** [comment]
Hi @cofinoa 

your effort is really appreciated. If you could provide a patch it would be great.

Thanks again,
Marco

**Comment by mcucchi9** [comment]
@cofinoa with respect to https://github.com/Unidata/netcdf-java/issues/997#issuecomment-1100085618, how did you manage to dump the encoded variables with toolsUI 5.5.2?

**Comment by cofinoa** [comment]
@mcucchi9 
in the ToolsUI app, go to the `Iosp->HDF5->Netcdf4-JNI` tab and open the netcdf file. This will use the netcdf4-jni IOSP for reading the netcdf file. 

![image](https://user-images.githubusercontent.com/5098603/165319295-d8f34230-178f-4438-b6f2-3d89416e6e22.png)

**Comment by cofinoa** [comment]
@mcucchi9 

I have created a new jar with a partial fix. You can download from: 
http://sanfernando72.cofino.es/owncloud/index.php/s/eJUAjqCLxlQSRbc/download
and run the toolsUI with:
```
java -Djna.library.path=$NETCDF_LIB_PATH -jar toolsUI-5.5.2-FIX997.jar -nj22Config ../nj22Config.xml
```
with the `$NETCDF_LIB_PATH` pointing at the same LIB directory as in the `nj22Config.xml`

 Please test it and let me know. 

P.S: As I have mention, this is just a quick, dirty and partial fix for the actual issue.

**Comment by mcucchi9** [comment]
@cofinoa 

Thanks, the jar you provided is working indeed!

A couple of questions:

1. Have you got any idea on the required time for the release of a public fixed version?
2. Would the fix also impact other netCDF-Java based softwares, such as Panoply (which, as I mentioned in my first comment, presents similar problems)?

Thanks again for your efforts!

**Comment by cofinoa** [comment]
@mcucchi9

The time for a public fixed release it's something that depends on Unidata developers. 

I can propose a fix and the corresponding PR, but depends on them to accept the fix. At the moment I have just track-down the actual issue and the fix will require some effort. Please If you want a solution for a an important project at BOPEN, please contact me, you can find my contact e-mail at Github profile: @cofinoa 

With respect to the second question, the panoply it's suffering from the same issue, and it can be fixed (hope so) if the netcdf-java library jar embeded, it's replaced by a new one ..... but needs a better understanding.

**Comment by haileyajohnson** [comment]
@cofinoa we'd be really grateful if you made a pull request, or could share with us what you did for you fix. Thanks for looking at this!

**Comment by lesserwhirls** [comment]
It looks like the [RuntimeConfigParser](https://github.com/Unidata/netcdf-java/blob/3cac961e4054b7bd27905ee8bfd6221564f3c831/cdm/core/src/main/java/ucar/nc2/util/xml/RuntimeConfigParser.java#L296) uses the "old" API to load the IOSP when reading is enabled, so reading should work if you tell the config file to use the c library for reading and you open the file with `NetcdfFile.Open`. With the new API (`NetcdfFiles.open`), that is not the case. The fix might be as simple as adding a call to `NetcdfFiles.registerIOProvider(nc4IospClassName);` alongside the existing `NetcdfFile.registerIOProvider(nc4IospClassName);` call in the runtime config parser class.

**Comment by cofinoa** [comment]
I have opened a PR. 

The `setLibraryAndPath` method from `Nc4Iosp` has been deprecated and splited into `NetcdfClibrary`. 

Then the library path and name, needs to be made using the `setLibraryAndPath` from `NetcdfClibrary`, and register as IOSP reader using the `Nc4Iosp`. 

I have just tested and toolsUI loads and applies the `nj22config` elements, but I have not tested all the corner cases. 

@lesserwhirls, what worries to me is that [Nc4Iosp](https://github.com/Unidata/netcdf-java/blob/c2293e9ea4d1752beec6a15bc88d1c3317e2fce6/netcdf4/src/main/java/ucar/nc2/jni/netcdf/Nc4Iosp.java#L61) invokes, statically, the `NetcdfCLibrary` static method .... and I'm not sure if this would produce some issues in the future.
